dev/src/bundle.ts

```diff
@@ -0,0 +1,225 @@
+
+  // The latest read time among all bundled documents and queries.
+  private latestReadTime: Timestamp = Timestamp.fromProto({
+    seconds: 0,
+    nanos: 0,
+  });
+
+  constructor(private id: string) {}
+
+  add(doc: DocumentReference): BundleBuilder;
+  add(docSnap: DocumentSnapshot): BundleBuilder;
+  add(queryName: string, querySnap: QuerySnapshot): BundleBuilder;
+  add(queryName: string, query: Query): BundleBuilder;
+  add(
+    docOrName: DocumentReference | DocumentSnapshot | string,
+    query?: QuerySnapshot | Query
+  ) {
+    if (docOrName instanceof DocumentReference) {
+      this.docRefs.push(docOrName);
+    } else if (docOrName instanceof DocumentSnapshot) {
+      this.docSnaps.push(docOrName);
+    } else if (query instanceof Query) {
+      this.queries.set(docOrName, query);
+    } else if (query instanceof QuerySnapshot) {
+      this.querySnaps.set(docOrName, query);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private pushLengthPrefixedString(readable: Readable, payload: string) {
+    const buffer = Buffer.from(payload, 'utf-8');
+    const lengthBuffer = Buffer.from(buffer.length.toString());
+
+    readable.push(Buffer.concat([lengthBuffer, buffer]));
+  }
+
+  stream(): NodeJS.ReadableStream {
+    assert(
+      this.docRefs.length +
+        this.docSnaps.length +
+        this.queries.size +
+        this.querySnaps.size >
+        0,
+      'Nothing is added to the bundle.'
+    );
+    for (const snap of this.docSnaps) {
+      this.addBundledDocument(snap);
+    }
+
+    for (const [name, snap] of Array.from(this.querySnaps)) {
+      if (this.namedQueries.has(name)) {
+        throw new Error(`Query name conflict: ${name}`);
+      }
+
+      this.addNamedQuery(name, snap);
+    }
+
+    const promises: Array<Promise<void>> = [];
+    for (const doc of this.docRefs) {
+      promises.push(
```

Comment:
It is surprising to me that the `stream()` API fetches the documents from the backend. I think we should maybe revisit the API here and change add() to only take DocumentSnapshots or QuerySnapshots (and take the opportunity to split the methods into two instead of overloading them). 

If we want to keep the ability to add DocumentReferences, then `add()` should be async and do perform the document retrieval.

Reply:
Done, no more references or queries, only snapshots.